CnCNet/Net/PeerToPeer/PeerToPeerUtil.cs

```diff
@@ -1,135 +1,121 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Text;
+﻿namespace CnCNetServer;
+
 using System.Net;
 using System.Net.Sockets;
-using System.Threading.Tasks;
-using System.Timers;
-using System.Linq;
+using Microsoft.Extensions.Logging;
 
-namespace CnCNet.Net.PeerToPeer
+internal sealed class PeerToPeerUtil : IDisposable
 {
-    class PeerToPeerUtil
+    private const int CounterResetInterval = 60 * 1000; // Reset counter every X ms
+    private const int MaxRequestsPerIp = 20; // Max requests during one CounterResetInterval period
+    private const int MaxConnectionsGlobal = 5000; // Max amount of different ips sending requests during one CounterResetInterval period
+    private const int StunId = 26262;
+
+    private readonly Dictionary<int, int> connectionCounter = new(MaxConnectionsGlobal);
+    private readonly System.Timers.Timer connectionCounterTimer = new(CounterResetInterval);
+    private readonly byte[] sendBuffer = new byte[40];
+    private readonly SemaphoreSlim connectionCounterSemaphoreSlim = new(1, 1);
+    private readonly ILogger logger;
+
+    public PeerToPeerUtil(ILogger<PeerToPeerUtil> logger)
     {
-        const int COUNTER_RESET_INTERVAL = 60 * 1000; // Reset counter every X ms
-        const int MAX_REQUESTS_PER_IP = 20; // Max requests during one COUNTER_RESET_INTERVAL period
-        const int MAX_CONNECTIONS_GLOBAL = 5000; // Max amount of different ips sending requests during one COUNTER_RESET_INTERVAL period
-        const int STUN_ID = 26262;
+        this.logger = logger;
+        new Random().NextBytes(sendBuffer);
+        Array.Copy(BitConverter.GetBytes(IPAddress.HostToNetworkOrder((short)StunId)), 0, sendBuffer, 6, 2);
+    }
 
-        int ListenPort;
-        UdpClient Client;
-        Task ReceiveTask;
-        Dictionary<int, int> ConnectionCounter = new Dictionary<int, int>(MAX_CONNECTIONS_GLOBAL);
-        Timer ConnectionCounterTimer = new Timer(COUNTER_RESET_INTERVAL);
-        object _lock = new object();
-        byte[] SendBuffer = new byte[40];
+    public Task StartAsync(int listenPort, CancellationToken cancellationToken)
+    {
+        connectionCounterTimer.Elapsed += (_, _) => _ = ResetConnectionCounterAsync(cancellationToken);
```

Comment:
This particular line asks for an explanation on what's happening in that lambda function.

Reply:
This is equivalent to:
```C#
+= (object? sender, ElapsedEventArgs args) => ResetConnectionCounterAsync(cancellationToken);
```
This however leaves the 2 parameters `sender` and `args` unused. To indicate that this is intentional and we did not forget to pass them to `ResetConnectionCounterAsync` the discard symbol is used.
This makes us end up with:
```C#
+= (_, _) => ResetConnectionCounterAsync(cancellationToken);
```
Both this and the previous code generate a warning however: [Compiler Warning (level 1) CS4014](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs4014) due to the fact that `ResetConnectionCounterAsync` is an async method and we do not await it. You cannot await a method in an event handler however meaning an event handler will never wait for an async method to finish and rather return before the method has run to completion. This is typically fine however for most events.
To indicate that we are fine with this we discard the `Task` return value of the method:
```C#
+= (_, _) => _ = ResetConnectionCounterAsync(cancellationToken)
```
The warning CS4014 will consequently not be raised.